### ISSUE TITLE
php-cs-fixer vs Windows 10: executablePath not found

### ISSUE THREAD

**Issue by kreinholz** [susanBuck/e15-spring20#9]
I followed the setup directions [here](https://hesweb.dev/e15/notes/php/code-style#vscode-php-cs-fixer) to configure PHP CS Fixer after installing the extension on VS Code on Windows 10.

I set the exectuablePath as follows:

``` js
"php-cs-fixer.executablePath": "c:/Users/Kevin Reinholz/.vscode/extensions/junstyle.php-cs-fixer-0.1.129/php-cs-fixer.phar",
```
Yet when I try to save a .php file, I receive the following error:

> PHP CS Fixer: executablePath not found, please check your settings. It will set to built-in php-cs-fixer.phar. Try again!

And settings.json then automatically reverts to the nonworking default:

```js
"php-cs-fixer.executablePath": "${extensionPath}\\php-cs-fixer.phar",
```

I checked in Cmder and my path is correct--the directory and the file both exist:

```
C:\Users\Kevin Reinholz\.vscode\extensions\junstyle.php-cs-fixer-0.1.129  (php-cs-fixer@0.1.129)
λ ls
beautifyHtml.js  download-phar.js  logo.png      package.json       README.md
CHANGELOG.md     extension.js      node_modules  php-cs-fixer.phar
```

I'm going to return to **project1** for now, but I would like to get this extension working...I can't for the life of me figure out what I've done wrong in settings.json. Thanks!

**Comment by RGPhD** [comment]
Hi Kevin @kreinholz ,
My file path for the php-cs-fixer on one of my Windows 10 PCs is ```C:\Users\RGPhD\.vscode\extensions\junstyle.php-cs-fixer-0.1.129\php-cs-fixer.phar```.  It looks a little bit different than yours ```C:\Users\Kevin Reinholz\.vscode\extensions\junstyle.php-cs-fixer-0.1.129  (php-cs-fixer@0.1.129)``` ... but it works.  I noticed in the instructions by @susanBuck the path has forward slashes and our paths have the Windows back slashes.  I will check on my other PC to see if I saved the same new lines in the json file and if it has the same file path.

Professor Buck also notes that *The auto-formatter won’t run if you have any PHP syntax/parse errors; these have to be cleared first.*

Wish I could be of more help,
Robelyn

**Comment by kreinholz** [comment]
@RGPhD thank you for your reply. I am convinced the issue is the space in my user profile name on Windows, but I'm not going to create a new profile just to run this extension (and frankly there should be a sane way to escape a space in a .json file path).

I've tried numerous variations, such as:

```
"php-cs-fixer.executablePath": "%USERPROFILE%\\.vscode\\extensions\\junstyle.php-cs-fixer-0.1.129\\php-cs-fixer.phar",
```

Or:

```
"php-cs-fixer.executablePath": "C:\\Users\\Kevin Reinholz\\.vscode\\extensions\\junstyle.php-cs-fixer-0.1.129\\php-cs-fixer.phar",
```

Or the above with putting 'Kevin Reinholz' inside single quotes. Always with the same result: "executablePath not found"...

I've tried uninstalling and reinstalling junstyle's php-cs-fixer extension, same story. I'd like to get it working, but it doesn't seem to like Windows very much, especially when there's a dastardly space in the user profile path.

Of further frustration, other extensions to VS Code work just fine, so the executablePath fragility seems to be limited to this one...

**Comment by RGPhD** [comment]
Hi @kreinholz Kevin,

That makes total sense.  I do not have a space in either of my PC usernames.

Robelyn

**Comment by susanBuck** [comment]
Hi @kreinholz -

Try this path:

```
c:/Users/KEVINR~1/.vscode/extensions/junstyle.php-cs-fixer-0.1.129/php-cs-fixer.phar
```

It substitutes `Kevin Reinholz` for the DOS shortname equivalent of `KEVINR~1`.

(For more on what the DOS shortname is and how to find it, [check out this post...](https://stackoverflow.com/a/59549445/59479).

If that doesn't work, let me know. I'll do some experiments on my Windows machine with a multi-word user account.

**Comment by rsiu01** [comment]
I had the same problem as Kevin. I found a [solution](https://github.com/junstyle/vscode-php-cs-fixer/issues/25#issuecomment-357416926) to this issue online, which requires installing php cs fixer via composer.

**Comment by kreinholz** [comment]
Alas, changing my path as @susanBuck suggested did not clear the error, as much as good old DOS 8.3 file naming conventions seemed sure to work!

I found a "workaround": by installing PHP and adding the following line, I at least seem to get _some_ PHP auto-formatting:

`"php.validate.executablePath": "C:\\php\\php.exe"`

Not sure if that's a good answer or not, but for now errors are silenced and my code seems cleaner from a stylistic standpoint when I save...

**Comment by susanBuck** [comment]
@kreinholz Glad you found a solution that worked so you can move forward. : )

I'll post again later after I dig some more into this - I'm still curious to see if we can figure out a way to have it work with multi-word usernames, as I certainly anticipate future students running into the same problem.

**Comment by susanBuck** [comment]
I had some time to dig into this more -

One solution I found was to move `php-cs-fixer.phar` from `c:/Users/Your Username/.vscode/extensions/junstyle.php-cs-fixer-0.1.129/` to `c:\`, circumventing the need to have the username in the path. So the setting simply became:

```
"php-cs-fixer.executablePath": "c:\php-cs-fixer.phar",
```

That did the trick.

But I was still curious about why the default path with the `${extensionPath}` variable didn't just work, as that is the most straightforward solution and OS-independent.

I did a bunch of tests and I think I was able to nail down a pattern to get it to work:

To set up my tests, I first uninstalled the `vscode-php-cs-fixer` extension.

Then I installed it, not making any setting changes.

I then saved a PHP file to trigger the formatter, and got the error:

> PHP CS Fixer: executablePath not found, please check your settings. It will set to built-in php-cs-fixer.phar. Try again!

I opened my settings and confirmed the path was as follows:

```
"php-cs-fixer.executablePath": "${extensionPath}/php-cs-fixer.phar",
```

Finally, I just **restarted VSCode**, and voila, it was working.

I did this procedure a couple times on both my Mac and Windows machine and it worked each time, so I'm hoping it's as simple as that - just restart VSCode after installing the plugin.

If anyone has a chance to test this procedure on their and and see if it lets `${extensionPath}/php-cs-fixer.phar` working for them - please let me know! : ) I'd love to be able to simplify the notes/instructions to this without having to have students mess with OS-specific paths.

**Comment by kreinholz** [comment]
@susanBuck, as odd as it seems (I'm convinced I restarted VS Code multiple times after installing php-cs-fixer), removing my workaround line:

`"php.validate.executablePath": "C:\\php\\php.exe"`

And restarting with the default:

`"php-cs-fixer.executablePath": "${extensionPath}/php-cs-fixer.phar",`

Didn't cough up any further errors when saving a php file, so for now at least I consider this issue resolved! If it resurfaces, I'll play around with it some more and comment on it.

Also, after reviewing the Week 04 materials, I realized my workaround was a bad one--I should've done the following instead:

`"php.validate.executablePath": "C:\\xampp\\php\\php.exe"`

Rather than installing a second, competing version of php (I've since uninstalled that "extra" php installation).

That said, my workaround appears to be unnecessary, so thank you very much for the time and experimentation sent trying to duplicate and resolve my error!
